src/datta_lab_to_nwb/markowitz_gillis_nature_2023/optogeneticinterface.py

```diff
@@ -0,0 +1,88 @@
+"""Primary class for converting fiber photometry data (dLight fluorescence)."""
+# Standard Scientific Python
+import pandas as pd
+import numpy as np
+
+# NWB Ecosystem
+from pynwb.file import NWBFile
+from pynwb.core import DynamicTableRegion
+from pynwb.ophys import RoiResponseSeries
+from pynwb.ogen import OptogeneticSeries
+from neuroconv.basedatainterface import BaseDataInterface
+from neuroconv.utils import load_dict_from_file
+from neuroconv.tools import nwb_helpers
+from hdmf.backends.hdf5.h5_utils import H5DataIO
+
+# Local
+
+
+class OptogeneticInterface(BaseDataInterface):
+    """Optogenetic interface for markowitz_gillis_nature_2023 conversion"""
+
+    def __init__(self, file_path: str, session_uuid: str, metadata_path: str):
+        # This should load the data lazily and prepare variables you need
+        columns = (
+            "uuid",
+            "feedback_status",
+            "stim_duration",
+            "stim_frequency",
+            "power",
+            "area",
+            "timestamp",
+        )
+        super().__init__(
+            file_path=file_path,
+            session_uuid=session_uuid,
+            columns=columns,
+            metadata_path=metadata_path,
+        )
+
+    def get_metadata(self) -> dict:
+        metadata = super().get_metadata()
+        return metadata
+
+    def get_metadata_schema(self) -> dict:
+        metadata_schema = super().get_metadata_schema()
+        return metadata_schema
+
+    def run_conversion(self, nwbfile: NWBFile, metadata: dict):
+        """Run conversion of data from the source file into the nwbfile."""
+        session_df = pd.read_parquet(
+            self.source_data["file_path"],
+            columns=self.source_data["columns"],
+            filters=[("uuid", "==", self.source_data["session_uuid"])],
+        )
+        location = session_df["area"].unique()[0]
+        stim_duration = session_df["stim_duration"].unique()[0]
+        fs = 30.0
+        stim_duration_index = int(stim_duration * fs)
+        power_mW = session_df["power"].unique()[0]
+        power_W = power_mW / 1000.0
+
+        device = nwbfile.create_device(
+            name="Opto Engine MRL-III-635",
+            description="Optogenetic stimulator (Opto Engine MRL-III-635; SKU: RD-635-00500-CWM-SD-03-LED-0)",
+            manufacturer="Opto Engine LLC",
+        )
+        ogen_site = nwbfile.create_ogen_site(
+            name="OptogeneticStimulusSite",
+            device=device,
+            description="Optogenetic stimulus site",
+            excitation_lambda=635.0,
+            location=location,
+        )
+        # Reconstruct optogenetic series from feedback status
+        feedback_status_cts = session_df.feedback_status.to_numpy()
+        feedback_is_on_index = np.where(feedback_status_cts == 1)[0]
+        for index in feedback_is_on_index:
+            feedback_status_cts[index : index + stim_duration_index] = 1
+        feedback_status_cts[feedback_status_cts == -1] = 0
```

Comment:
It should be sufficient to only write the `data` and `timestamps` of non-zero power signal

Reply:
Is that the normal approach?

Reply:
Not that there is much history to establish as 'normal', but for the recent ogen implementations, yes. Though to clarify, I don't mean all zeroes, just contiguous entries; we need to communicate the 'off' part of the signal to indicate durations after all. So the pattern would be `data=(..., power_value_during_pulse, 0, power_value_during_next_pulse, 0)`, `timestamps=(..., timing_of_on_pulse, timing_of_off_pulse, timing_of_next_on_pulse, timing_of_next_off_pulse, ...)`

It also just makes sense from a data engineering standpoint - most of the series in the file is a vector of zeroes - why take up all that space when you can communicate the same thing in a more condensed way? The important details are the (i) precise timestamps of the stimulation, (ii) strength of the power during the pulse, and (iii) shape and duration of the pulse (duration can best estimated from (i) and your shape is only a simple square)

Reply:
Ok, I've implemented a solution in this style.